src/framework/uicomponents/view/buttonboxmodel.cpp

```diff
@@ -0,0 +1,155 @@
+
+#include "log.h"
+
+using namespace mu::uicomponents;
+
+ButtonBoxModel::ButtonBoxModel(QObject* parent)
+    : QObject(parent), m_buttonsItems(this)
+{
+}
+
+QList<int> ButtonBoxModel::load()
+{
+    QList<int> result;
+    std::unordered_map <int, std::vector <LayoutButton*> > sortedButtons;
+    int maxCustomRole = static_cast<int>(ButtonRole::CustomRole);
+
+    QList<QQuickItem*> buttonsItems = m_buttonsItems.list();
+    for (const QQuickItem* item : buttonsItems) {
+        QVariant buttonTypeVar = item->property("buttonId");
+        int type = static_cast<int>(ButtonType::CustomButton);
+        LayoutButton* button = nullptr;
+        if (buttonTypeVar.isValid()) {
+            type = buttonTypeVar.toInt();
+            if (type >= static_cast<int>(ButtonType::CustomButton)) {
+                button = layoutButton(item);
+            } else {
+                button = m_layoutButtons[static_cast<ButtonType>(type)];
+            }
+        } else {
+            button = m_layoutButtons[static_cast<ButtonType>(type)];
+        }
+
+        ButtonRole role = button->buttonRole;
+        if (role >= ButtonRole::CustomRole) {
+            maxCustomRole = std::max(maxCustomRole, static_cast<int>(role));
+        }
+
+        sortedButtons[role].push_back(button);
+    }
+
+    const std::vector<ButtonRole> currentLayout = chooseButtonLayoutType();
+
+    auto buttonsByRole = [&sortedButtons, maxCustomRole](ButtonRole role) -> std::vector<LayoutButton*> {
+        bool isCustom = role == ButtonRole::CustomRole;
+        if (!isCustom) {
+            if (contains(sortedButtons, static_cast<int>(role))) {
+                return sortedButtons[role];
+            }
+
+            return {};
+        }
+
+        std::vector<LayoutButton*> buttons;
+        for (int role = static_cast<int>(ButtonRole::CustomRole); role <= maxCustomRole; ++role) {
+            for (LayoutButton* button : sortedButtons[role]) {
+                buttons.push_back(button);
+            }
+        }
+
+        return buttons;
+    };
+
+    for (ButtonRole buttonRole : currentLayout) {
+        std::vector<LayoutButton*> buttons = buttonsByRole(buttonRole);
+
+        for (LayoutButton* button : buttons) {
+            result << button->buttonType;
+        }
+    }
+
+    return result;
+}
+
+void ButtonBoxModel::setButtons(const QVariantList& buttons)
+{
+    for (const QVariant& buttonTypeVar : buttons) {
+        LayoutButton* button = m_layoutButtons[ButtonType(buttonTypeVar.toInt())];
+        emit addButton(button->text, button->buttonType, button->buttonRole, button->isAccent, button->isLeftSide);
+    }
+}
+
+const std::vector <ButtonBoxModel::ButtonRole> ButtonBoxModel::chooseButtonLayoutType()
+{
+    size_t index = 0;
+    if (m_buttonLayout != ButtonLayout::UnknownLayout) {
+        index = static_cast<size_t>(m_buttonLayout);
+    } else {
+#if defined (Q_OS_OSX)
+        index = 1;
+#elif defined (Q_OS_LINUX) || defined (Q_OS_UNIX)
```

Comment:
Maybe we also need ` || defined(Q_OS_FREEBSD)` here (we also have to support Freebsd nowadays, since there was a community PR that added it). Or does freebsd use Windows style?